metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/execution/WorkflowExecutionMonitor.java

```diff
@@ -0,0 +1,335 @@
+ * this eventuality. Two of its methods are meant to be scheduled for a periodical run.
+ */
+public class WorkflowExecutionMonitor {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(WorkflowExecutionMonitor.class);
+
+  private static final String FAILSAFE_LOCK = "failsafeLock";
+
+  private final WorkflowExecutionDao workflowExecutionDao;
+  private final WorkflowExecutorManager workflowExecutorManager;
+  private final Duration failsafeLeniency;
+  private final RLock lock;
+
+  /** The currently running executions. **/
+  private Map<String, WorkflowExecutionEntry> currentExecutions = Collections.emptyMap();
+
+  /**
+   * Constructor the executor
+   *
+   * @param workflowExecutorManager {@link WorkflowExecutorManager}
+   * @param workflowExecutionDao {@link WorkflowExecutionDao}
+   * @param redissonClient {@link RedissonClient}
+   * @param failsafeLeniency The leniency given to executions to be idle.
+   */
+  public WorkflowExecutionMonitor(WorkflowExecutorManager workflowExecutorManager,
+      WorkflowExecutionDao workflowExecutionDao, RedissonClient redissonClient,
+      Duration failsafeLeniency) {
+    this.failsafeLeniency = failsafeLeniency;
+    this.workflowExecutionDao = workflowExecutionDao;
+    this.workflowExecutorManager = workflowExecutorManager;
+    this.lock = redissonClient.getFairLock(FAILSAFE_LOCK);
+  }
+
+  /**
+   * Makes a run to make an inventory of the last update times of all running tasks. This method is
+   * meant to run periodically.
+   */
+  public void performExecutionMonitoring() {
+    try {
+
+      // Lock for the duration of this scheduled task.
+      lock.lock();
+
+      // Update the current executions map (completely replace it).
+      updateCurrentExecutions();
+
+    } catch (RuntimeException e) {
+      LOGGER.warn("Exception thrown from Redis, monitoring thread continues", e);
+    } finally {
+      try {
+        lock.unlock();
+      } catch (RedisConnectionException e) {
+        LOGGER.warn("Cannot connect to unlock, monitoring thread continues", e);
+      }
+    }
+  }
+
+  List<WorkflowExecution> updateCurrentExecutions() {
+
+    // Check lock
+    if (!lock.isHeldByCurrentThread()) {
+      throw new IllegalStateException();
+    }
+
+    // Get all workflow executions that are currently running
+    final List<WorkflowExecution> allRunningWorkflowExecutions =
+        getWorkflowExecutionsWithStatus(WorkflowStatus.RUNNING);
+
+    // Go by all running executions and compare them with the data we already have.
+    final Map<String, WorkflowExecutionEntry> newExecutions = new HashMap<>();
+    for (WorkflowExecution execution : allRunningWorkflowExecutions) {
+      final WorkflowExecutionEntry currentEntry = getEntry(execution);
+      final WorkflowExecutionEntry newEntry;
+      if (currentEntry != null && currentEntry.updateTimeValueIsEqual(execution.getUpdatedDate())) {
```

Comment:
No, it is because I need the lastValueChange property of the entry to remain unchanged. So there has to be an if statement anyway (to decide on the lastValueChange: it is only updated when the update time changed). 

Since I preferred to keep WorkflowExecutionEntry an immutable class, this is the solution I chose.

Reply:
You are right I missed that, sorry. Perhaps add some comment there?

Reply:
I added a bit to the comments.